firewall-km.c

```diff
@@ -143,8 +186,182 @@ struct file_operations firewall_fops = {
+                    continue;
+                }
+            }
+            if(rule_found->destination_ip==0){
+                //...............destination ip not specified => match............
+            } else {
+                if(!check_ip_integrity(dest_ip,rule_found->destination_ip,rule_found->destination_netmask)){
+                    printk(KERN_INFO "rule %d not matched: dest ip mismatch\n",i);
+                    continue;
+                }
+            }
+            if(rule_found->source_port==0){
+                //..............source port not specified => match..............
+            } else if (src_port!=rule_found->source_port){
+                printk(KERN_INFO "rule %d not matched: source port mismatch\n",i);
+                continue;
+            }
+            if(rule_found->destination_port==0){
+                //..........destination port not specified => match............
+            } else if(dest_port!=rule_found->destination_port){
+                printk(KERN_INFO "rule %d not matched: destination port mismatch\n",i);
+                continue;
+            }
+                //match found check action to be taken
+            if(rule_found->action==0){
+                printk(KERN_INFO "Match found: %d, packet is dropped\n",i);
+                return NF_DROP;
+            } else {
+                printk(KERN_INFO "Match found: %d, packet accepted\n",i);
+                return NF_ACCEPT;
+            }
+        }
+    }
+    printk(KERN_INFO "No matching found, packet is accepted\n");
+    return NF_ACCEPT;
+}
+
+unsigned int hook_func_out(void *priv, struct sk_buff *skb, const struct nf_hook_state *state) {
+    struct iphdr *ip_hdr = (struct iphdr *)skb_network_header(skb);
+    struct udphdr *udp_hdr;
+    struct tcphdr *tcp_hdr;
+
+    struct list_head *p;
+    struct mf_rule *rule_found;
+
+    char src_ip_str[16], dest_ip_str[16];
+    int i=0;
+
+    unsigned int src_ip = (unsigned int) ip_hdr->saddr;
+    unsigned int dest_ip = (unsigned int) ip_hdr->daddr;
+    unsigned int src_port,dest_port;
+
+    if (ip_hdr->protocol==17){
+        udp_hdr = (struct udphdr *)skb_transport_header(skb);
+        src_port = (unsigned int)ntohs(udp_hdr->source);
+        dest_port = (unsigned int)ntohs(udp_hdr->dest);
+    } else if (ip_hdr->protocol==6){
+        tcp_hdr = (struct tcphdr *)skb_transport_header(skb);
+        src_port = (unsigned int)ntohs(tcp_hdr->source);
+        dest_port = (unsigned int)ntohs(tcp_hdr->dest);
+    }
+
+    ip_hl_to_str(ntohl(src_ip), src_ip_str);
+    ip_hl_to_str(ntohl(dest_ip), dest_ip_str);
+
+    //printk(KERN_INFO "OUT packet info src ip: %u = %s, src port: %u; dest ip: %u = %s, dest port: %u; protocol: %u\n",src_ip,src_ip_str,src_port,dest_ip,dest_ip_str,dest_port,ip_hdr->protocol);
+
+    list_for_each(p,&policy_list.list){
+        i++;
+        rule_found = list_entry(p, struct mf_rule, list);
+        if(rule_found->inbound_outbound!=1){
```

Comment:
Dont't check it like this, make a similar struct to mf_rule out of the info extracted from the packages and make a boolean function that returns if we should let the package pass or not